RFCs/FS-1060-nullable-reference-types.md

```diff
@@ -0,0 +1,1033 @@
+
+The desired outcome over time is that significantly less `NullReferenceException`s are produced by code at runtime when itneroperating with C# libraries and other F# components that declare reference types as capable of having `null` as a proper value. This feature should allow F# programmers to more safely eject `null` from their concerns when interoperating with other .NET components, and make it more explicit when `null` could be a problem.
+
+## Principles
+
+The following principles guide all further considerations and the design of this feature.
+
+1. On the surface, F# is "almost" as simple to use as it is today. In practice, it must feel simpler due to nullability of types like `string` being explicit rather than implicit.
+2. The value for F# is primarily in flowing non-nullable reference types into F# code from .NET Libraries and from F# code into .NET libraries.
+3. Adding nullable annotations should not be part of routine F# programming.
+4. Nullability annotations/information is carefully suppressed and simplified in tooling (tooltips, FSI) to avoid extra information overload.
+5. F# users are typically concerned with this feature at the boundaries of their system so that they can flow non-null data into the inner parts of their system.
+6. The feature produces warnings by default, with different classes of warnings (nullable, non-nullable) offering opt-in/opt-out mechanisms.
+7. All existing F# projects compile with warnings turned off by default. Only new projects have warnings on by default. Compiling older F# code with newer F# code does not opt-in the older F# code.
+8. F# non-nullness is reasonably "strong and trustworthy" today for F#-defined types and routine F# coding. This includes the explicit use of `option` types to represent the absence of information. No compile-time guarantees today will be "lowered" to account for this feature.
+9. The F# compiler will strive to provide flow analysis such that common null-check scenarios guarantee null-safety when working with nullable reference types.
+10. This feature is useful for F# in other contexts, such as interoperation with `string` in JavaScript via [Fable](http://fable.io/).
+11. F# programmers may be able to start to phase out their use of `[<AllowNullLiteral>]`, `Unchecked.defaultof<_>`, and the `null` type constraint when they need ad-hoc nullability for reference types declared in F#.
+12. Syntax for the feature feels "baked in" to the type system, and should not be horribly unpleasant to use.
+
+## Detailed design
+[design]: #detailed-design
+
+### .NET Metadata
+
+To remain backwards-compatible with existing codebases and interoperate with other .NET languages that support nullability as a concept, F# will emit and respect .NET metadata that concers to the following:
+
+* If a type is a nullable reference type
+* If an assembly has nullability as a concept (regardless of what it was compiled with)
```

Comment:
We should discuss this. Latest C# design does not include such an assembly-level marker. We may need to revive it if it's needed.